In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report



---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame

csv_path = Path("Resources/lending_data.csv")
df = pd.read_csv(csv_path)

# Review the DataFrame
print(df.head())

   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0  


### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features


# Separate the y variable, the labels
y = df["loan_status"]

# Separate the X variable, the features
X = df.drop(columns="loan_status")

In [4]:
# Review the y variable Series
print(y.head())

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64


In [5]:
# Review the X variable DataFrame
print(X.head())

   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  


### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
print(y.value_counts())

0    75036
1     2500
Name: loan_status, dtype: int64


### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [28]:
# Import the LogisticRegression module from SKLearn
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
logistic_regression_model.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [29]:
# Make a prediction using the testing data
y_pred = logistic_regression_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [30]:
# Print the balanced_accuracy score of the model
accuracy_score_result = balanced_accuracy_score(y_test, y_pred)
print(f"Accuracy Score: {accuracy_score_result}")

Accuracy Score: 0.9520479254722232


In [31]:
# Generate a confusion matrix for the model
confusion_matrix_result = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(confusion_matrix_result)

Confusion Matrix:
[[18663   102]
 [   56   563]]


In [32]:
# Print the classification report for the model
classification_report_result = classification_report(y_test, y_pred)
print("Classification Report:")
print(classification_report_result)

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model has an accuracy score of 0.952 which is quite good.
The confusion matrix shows that the model has made 18663 correct predictions for class 0 (healthy loans) and 563 correct predictions for class 1 (high-risk loans). However, there are 102 false positives and 56 false negatives.
The classification report provides more insight into the model's performance:
For class 0 (healthy loans): The model has a high precision of 1.00, a recall of 0.99, and an F1-score of 1.00.
For class 1 (high-risk loans): The model has a precision of 0.85, a recall of 0.91, and an F1-score of 0.88.
In conclusion, the logistic regression model predicts both class 0 (healthy loans) and class 1 (high-risk loans) labels quite well. The model performs exceptionally well for healthy loans, with almost perfect precision, recall, and F1-score. For high-risk loans, the model has a slightly lower precision, recall, and F1-score, but they are still reasonably good. The model can be considered useful for predicting creditworthiness, but it's essential to be cautious when dealing with high-risk loans, as the model's performance is not as strong as it is for healthy loans.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [33]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = ros.fit_resample(X_train, y_train)

In [34]:
# Count the distinct values of the resampled labels data
print(y_resampled.value_counts())

0    56271
1    56271
Name: loan_status, dtype: int64


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [35]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_resampled = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
logistic_regression_resampled.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
y_pred_resampled = logistic_regression_resampled.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [36]:
# Print the balanced_accuracy score of the model 
accuracy_score_resampled = balanced_accuracy_score(y_test, y_pred_resampled)
print(f"Accuracy Score (Resampled): {accuracy_score_resampled}")

Accuracy Score (Resampled): 0.9936781215845847


In [37]:
# Generate a confusion matrix for the model
confusion_matrix_resampled = confusion_matrix(y_test, y_pred_resampled)
print("Confusion Matrix (Resampled):")
print(confusion_matrix_resampled)

Confusion Matrix (Resampled):
[[18649   116]
 [    4   615]]


In [39]:
# Print the classification report for the model
classification_report_result = classification_report(y_test, y_pred)
print("Classification Report:")
print(classification_report_result)

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model fit with oversampled data has an accuracy score of 0.9937, which is excellent. The confusion matrix shows that the model has made 18649 correct predictions for class 0 (healthy loans) and 615 correct predictions for class 1 (high-risk loans). However, there are 116 false positives and only 4 false negatives. The classification report provides more insight into the model's performance:

For class 0 (healthy loans): The model has a high precision of 1.00, a recall of 0.99, and an F1-score of 1.00.
For class 1 (high-risk loans): The model has a precision of 0.85, a recall of 0.91, and an F1-score of 0.88.

In conclusion, the logistic regression model fit with oversampled data predicts both class 0 (healthy loans) and class 1 (high-risk loans) labels very well. The model performs exceptionally well for healthy loans, with almost perfect precision, recall, and F1-score. For high-risk loans, the model has a slightly lower precision, recall, and F1-score, but they are still reasonably good. The model can be considered useful for predicting creditworthiness, but it's essential to be cautious when dealing with high-risk loans, as the model's performance is not as strong as it is for healthy loans. However, it's worth noting that the oversampled model has a significantly reduced number of false negatives compared to the original model, making it better at identifying high-risk loans.